crnk-jpa/src/main/java/io/crnk/jpa/JpaEntityRepositoryBase.java

```diff
@@ -185,6 +184,10 @@ protected boolean fetchRelations() {
         // id may differ from primary key
         ResourceField idField = getIdField();
         I id = (I) idField.getAccessor().getValue(resource);
+        // id could have been created during persist
+        if (id == null && pk != null && idField.getElementType().isAssignableFrom(pk.getClass())) {
```

Comment:
oh, I overlooked this one, not related to flush after all. it is a bit tricky in this area. 

I think a safer method would be to check whether primaryKeyAttribute.getName() (on super type) matched the idField.getUnderlyingName. This casae we are safe and can just make use of "pk". If it does not match, the should maybe call a method "getIdFromEntity". This may be overriden by sub types, but by default I would call a mapper.map(entity) and the fetch the ID with getIdField(). not super efficient but correct and ok-ish for a safe method, the effort is spend in other places.

Reply:
it is also not 100% safe if a mapper would have the id of mapping fields with the same name to different other fields, but that seem not to sane. But for the sake of correctness it could make sense to move the entire getIdFromEntity logic into a dedicated method to make it customizable/overridable.

Reply:
I like the idea of `getIdFromEntity` method.
A few thoughts:
- The precondition check for `pk != null` should not be needed. `pk` cannot be null after a successful call to `em.persist(entity)`. If pk was set, the entity is updated or created. If pk was null it can either be autogenerated or persist will fail. It is impossible to persist an entity without id. So the precondition check will never fail.
- I do not particularly like the idea of calling the mapper. I would have to provide a tuple to the `unmap`. The first entry would have to be the entity, but what about the others? The mapper could rely on the presence of calculated values, etc.

After all, I think my current checks are not too bad.
- The entity's pk is only used if the DTO's id is null
- The only use case is the create use case. In all other cases, the ID must be set.
- The primary key type is checked to be assignable to the resource ID type
- Additional check for matching property names added